tools/statplot.py

```diff
@@ -1,229 +1,305 @@
+                self.statplot.set_linewidth(2)
+                self.statplot.set_marker('None')
+            elif type == 'marker':
+                self.statplot.set_linestyle('None')
+                self.statplot.set_marker('d')
+                self.statplot.set_markeredgecolor('xkcd:black')
+                self.statplot.set_markerfacecolor('xkcd:light purple')
+                self.statplot.set_markersize(7)
+                self.statplot.set_markeredgewidth(0.3)
+            self.toolbar.update()
+        self.FormatAxis('x', xscale)
+        self.FormatAxis('y', yscale)
+        self.fig.canvas.draw()
+        self.fig.canvas.flush_events()
+
+    def PopulateCombo(self, action, prevIterX=None, prevIterY=None):
+        for entry in sorted(self.entries.Paths()):
+            self.xCombo.append_text(entry)
+            self.yCombo.append_text(entry)
+        if action == 'load':
+            if 'ElapsedTime' in self.entries.Paths():
+                iterX = self.xCombo.get_model(). \
+                    get_iter(sorted(self.entries.Paths()).index('ElapsedTime'))
+                self.xCombo.set_active_iter(iterX)
+                self.yCombo.set_active(0)
+            else:
+                self.xCombo.set_active(0)
+                self.yCombo.set_active(1)
+        elif action == 'reload':
+            self.xCombo.set_active_iter(prevIterX)
+            self.yCombo.set_active_iter(prevIterY)
+
+    def PopulateCompletion(self, text):
+        return [item for item in sorted(self.entries.Paths())
+                if item.startswith(text)]
+
+    def ReadData(self, statfile):
+        stats = []
+        for i, filename in enumerate(statfile):
+            failcount = 0
+            while failcount < 5:
+                try:
+                    stats.append(fluidity_tools.Stat(filename))
+                    break
+                except (TypeError, ValueError):
+                    time.sleep(0.2)
+                    failcount += 1
+            if failcount == 5:
+                raise Exception('Could not open %s' % filename)
+        if len(stats) == 1:
+            return stats[0]
+        else:
+            return fluidity_tools.JoinStat(*stats)
+
+    def RefreshData(self, statfile):
+        self.xCombo.disconnect_by_func(self.ComboChangedX)
+        self.yCombo.disconnect_by_func(self.ComboChangedY)
+        self.xCombo.remove_all()
+        self.yCombo.remove_all()
+        self.entries = self.ReadData(statfile)
+        self.PopulateCombo('reload', prevIterX=self.xCombo.get_active_iter(),
+                           prevIterY=self.yCombo.get_active_iter())
+        self.xCombo.connect('changed', self.ComboChangedX)
+        self.yCombo.connect('changed', self.ComboChangedY)
+
+    def ReleaseFocus(self, widget, event):
+        if event.keyval == Gdk.KEY_Escape:
+            self.canvas.grab_focus()
+
+
+window = StatplotWindow(args.statfile)
+window.connect('delete-event', Gtk.main_quit)
```

Comment:
Need to connect `destroy` here too to make the "q" keybinding work.

Reply:
From this [documentation](https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/Widget.html#Gtk.Widget.signals.delete_event), I am not entirely sure on the difference between the two. But judging from [this](https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/Widget.html#Gtk.Widget.destroy), I guess I will just connect `destroy`? Also, do you mean `q` is not working for you? And a last one on this: if I close the window using Alt+F4, I get the below error message in the Terminal. Any idea?
`Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/matplotlib/backends/backend_gtk3agg.py", line 40, in on_draw_event
    self._render_figure(w, h)
  File "/usr/lib/python2.7/dist-packages/matplotlib/backends/backend_gtk3agg.py", line 31, in _render_figure
    backend_agg.FigureCanvasAgg.draw(self)
  File "/usr/lib/python2.7/dist-packages/matplotlib/backends/backend_agg.py", line 433, in draw
    toolbar.set_cursor(toolbar._lastCursor)
  File "/usr/lib/python2.7/dist-packages/matplotlib/backends/backend_gtk3.py", line 494, in set_cursor
    self.canvas.get_property("window").set_cursor(cursord[cursor])
AttributeError: 'NoneType' object has no attribute 'set_cursor'`

Reply:
The `delete-event` is sent by the window manager (if you alt-F4, etc.). Pressing "q" is set up here to destroy the object, emitting the `destroy` signal instead. Indeed, for me, pressing "q" closed the window (because it was destroyde), but didn't quit the application. You should just need to connect `destroy` as well:

```suggestion
window.connect('delete-event', Gtk.main_quit)
window.connect('destroy', Gtk.main_quit)
```

I'm not really sure why you're getting those error messages, looks like the canvas isn't being cleaned up properly...